src/transformStateWithClones.js

```diff
@@ -6,8 +6,30 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let newState = { ...state };
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      newState = { ...newState, ...obj.extraData };
+
+      result.push(newState);
+    } else if (obj.type === 'removeProperties') {
+      for (const key of obj.keysToRemove) {
+        const { [key]: removed, ...rest } = newState;
+
+        newState = rest;
+      }
+      result.push({ ...newState });
+    } else if (obj.type === 'clear') {
+      newState = {};
+      result.push({ ...newState });
+    }
```

Comment:
Consider using a switch statement instead of multiple if-else conditions for better readability and maintainability, as suggested by the checklist.